(#349) android 10 gesture exclusion zone editor

K1rakishou opened the pull request:
Closes #349, #784

K1rakishou committed:
(#349) Implement base of the AdjustAndroid10GestureZonesController and the view

K1rakishou committed:
(#349) Current editable zone is now movable.
Use customly-implemented SafeRectF for calculations instead of the regular RectF which has it's top and bottom coordinated flipped.

K1rakishou committed:
(#349) Partial implementation of zones editing (left and right do work, top and bottom do not and there are couple of bugs)

K1rakishou committed:
(#349) Merge branch 'cheese-multi-feature' into (#349)-android-10-gesture-zone-editor

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanSettings.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/ExperimentalSettingsController.java
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou committed:
(#349) Merge branch 'cheese-multi-feature' into (#349)-android-10-gesture-zone-editor

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanSettings.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/ExperimentalSettingsController.java
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou committed:
(#349) Fix post-merge compilation errors

K1rakishou committed:
(#349) Merge branch '(#482)-targetsdk-29' into (#349)-android-10-gesture-zone-editor

K1rakishou committed:
(#349) Add ability to add Android 10 gestures exclusion zones

K1rakishou committed:
(#349) Add comments

K1rakishou committed:
(#349) Make the setting only be accessible on Android 10 and above

K1rakishou committed:
(#349) Merge branch 'cheese-multi-feature' into (#349)-android-10-gesture-zone-editor

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanSettings.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/ExperimentalSettingsController.java
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou committed:
(#349) Bug fixes and code review fixes

- Fix a bug where an exclusion zone would not get deep copied which would result in the original zone being mutated which is unexpected.

K1rakishou committed:
(#349) Update string

K1rakishou committed:
(#349) Merge branch 'cheese-multi-feature' into (#349)-android-10-gesture-zone-editor

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/FileCacheV2.kt
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/AppModule.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/ManagerModule.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/ChanLoaderManager.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/presenter/ThreadPresenter.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanSettings.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/MainSettingsController.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/layout/ReportProblemLayout.kt
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/utils/AndroidUtils.java
#	Kuroba/app/src/main/res/values/colors.xml
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou committed:
(#349) Post merge fixes

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/AdjustAndroid10GestureZonesView.kt:
```diff
@@ -0,0 +1,178 @@
+
+    @SuppressLint("ClickableViewAccessibility")
+    override fun onTouchEvent(event: MotionEvent): Boolean {
+        check(shown) { "View is not shown but onTouchEvent is called!" }
```
You should probably just ignore the touch event if the view isn't shown.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/AdjustAndroid10GestureZonesView.kt:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    override fun onTouchInProgress(x: Float, y: Float, dx: Float, dy: Float) {
+        check(shown) { "View is not shown but onMoving is called!" }
```
You should probably just ignore the touch event if the view isn't shown.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/AdjustAndroid10GestureZonesView.kt:
```diff
@@ -0,0 +1,178 @@
+
+    override fun onDraw(canvas: Canvas) {
+        super.onDraw(canvas)
+        check(shown) { "View is not shown but onDraw is called!" }
```
You should probably just ignore adding to the draw call if the view isn't shown. I don't think this situation would ever even occur if the view wasn't attached to a window.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/AdjustAndroid10GestureZonesView.kt:
```diff
@@ -0,0 +1,178 @@
+        onZoneAddedCallback = null
+        currentEditableZone = null
+        addedZones.clear()
+        shown = false
```
Might wanna add an invalidate call here.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/SafeRectF.kt:
```diff
@@ -0,0 +1,183 @@
+ *                                      |                                     |
+ * -------------------------------------+-------------------------------------+------------------------------------
+ * */
+class SafeRectF {
```
This class should rather be named "ScreenRectF" as that is its function. SafeRectF implies a RectF with additional safety checks, not changed functionality; the coordinate system would not have been implied to have changed.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/AdjustAndroid10GestureZonesController.kt:
```diff
@@ -0,0 +1,131 @@
+                View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
+        )
+
+        adjustZonesView.hide()
```
Is there a reason to call hide before show?

K1rakishou replied:
Just to reset the state.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/ExperimentalSettingsController.java:
```diff
@@ -1,25 +1,72 @@
+        //  When changing the following list don't forget to update indexes in LEFT_ZONES_INDEXES,
+        //  RIGHT_ZONES_INDEXES
+        //  !!!!!!!!!!!!!!!!!!!!!!!!!!
+        arrayAdapter.add(context.getString(R.string.setting_exclusion_zones_left_zone_portrait));
```
You know exactly what this comment points to.

K1rakishou replied:
???

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/Android10GesturesExclusionZonesHolder.kt:
```diff
@@ -0,0 +1,236 @@
+            return zones
+        }
+
+        if (!isAndroid10()) {
```
I don't think this check is necessary because you prevent even getting to this point through the settings controller.

K1rakishou replied:
Yeah, but you can export them with settings and then attempt to import on Android 9 and below. That's the point of the check.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/ExperimentalSettingsController.java:
```diff
@@ -77,9 +132,133 @@ private void setupConcurrentFileDownloadingChunksSetting(SettingsGroup group) {
+                R.string.setting_exclusion_zones_reset_zones,
+                R.string.setting_exclusion_zones_reset_zones_description,
+                (v) -> {
+                    if (!isAndroid10()) {
```
This check isn't necessary as you disable this option afterwards.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/ExperimentalSettingsController.java:
```diff
@@ -77,9 +132,133 @@ private void setupConcurrentFileDownloadingChunksSetting(SettingsGroup group) {
+    }
+
+    private void showZonesDialog() {
+        if (!isAndroid10()) {
```
This check isn't necessary as you disable this option afterwards.

K1rakishou replied:
It's there because onOptionClicked requires Android Q to be called.

Adamantcheese replied:
What I mean is that showZonesDialog is only called from the setting view, and that setting view is disabled if you're not on Android 10.

K1rakishou replied:
Yeah I know, but Android Studio doesn't like it. To fix that I will have to remove `@RequiresApi(Build.VERSION_CODES.Q)` from `AdjustAndroid10GestureZonesController` and I really don't want to do that because it helps preventing the controller getting called occasionally on a non Android 10 device.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/ExclusionZoneJson.kt:
```diff
@@ -0,0 +1,46 @@
+        val minScreenSize: Int,
+        @SerializedName("max_screen_size")
+        val maxScreenSize: Int,
+        @SerializedName("exclusion_zone")
```
Should be "exclusion_zones" as this is a List.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/EditableZone.kt:
```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+    private fun updateHandlePosition() {
```
This function and the one below it are essentially identical, probably a good idea to reuse the code if possible.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/Android10GesturesExclusionZonesHolder.kt:
```diff
@@ -0,0 +1,236 @@
+
+class Android10GesturesExclusionZonesHolder(
+        private val gson: Gson,
+        private val minScreenSize: Int,
```
Rather than passing these into the constructor, just get them as needed.

K1rakishou replied:
Yeah, no, let's leave like that. I tried moving that thing to other place and it started blowing up the app in endless loop right on the app start.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/AdjustAndroid10GestureZonesView.kt:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    override fun onTouchStart(x: Float, y: Float) {
+        val editableZone = checkNotNull(currentEditableZone) { "currentEditableZone is null" }
```
You should probably just ignore the touch event if the zone is null.

K1rakishou replied:
I use checkNotNull often because it does two things at once: ensures that currentEditableZone is not null and creates a new local non-null variable. Otherwise I will have to use "!!" which is considered a bad practice in Kotlin because it will just throw an NPE without any message.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/AdjustAndroid10GestureZonesView.kt:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    override fun onTouchEnd() {
+        val editableZone = checkNotNull(currentEditableZone) { "currentEditableZone is null" }
```
You should probably just ignore the touch event if the zone is null.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/Android10GesturesExclusionZonesHolder.kt:
```diff
@@ -0,0 +1,236 @@
+            return zones
+        }
+
+        if (json == EMPTY_JSON) {
```
You can collapse this check into the one above.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/Android10GesturesExclusionZonesHolder.kt:
```diff
@@ -0,0 +1,236 @@
+                    newExclusionZones += ExclusionZoneJson(
+                            screenOrientation = orientation,
+                            attachSide = zone.attachSide.id,
+                            left = zone.left.toFloat(),
```
You're saving these as Floats, but up above you're reading them as Ints. Did you mean to do that?

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/Android10GesturesExclusionZonesHolder.kt:
```diff
@@ -0,0 +1,236 @@
+        if (zones.size > 1) {
+            val zonesString = zones.joinToString(prefix = "[", postfix = "]", separator = ",")
+
+            throw IllegalStateException("More than one zone exists with the same orientation " +
```
You're throwing an exception here, but it looks like if you just let the function continue on then there would be no problems? Maybe change this to just be an error log statement?

K1rakishou replied:
Well this just means that something went wrong and I forgot to filter something somewhere because there should be only one zone per attach side. This exception is just a notification that something broke somewhere and needs to be investigated.

K1rakishou committed:
(#349) Code review fixes

K1rakishou committed:
(#349) Merge branch 'cheese-multi-feature' into (#349)-android-10-gesture-zone-editor

# Conflicts:
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou committed:
(#349) Merge branch 'cheese-multi-feature' into (#349)-android-10-gesture-zone-editor

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanSettings.java
#	Kuroba/app/src/main/res/values/strings.xml

K1rakishou committed:
(#349) Merge branch 'cheese-multi-feature' into (#349)-android-10-gesture-zone-editor

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/ChanSettings.java
#	Kuroba/app/src/main/res/values/strings.xml
